File: 5-1.py
#!/usr/bin/env python

import numpy
import re

class Map:
    def __init__(self):
        self.size = 1
        self.map = numpy.zeros((self.size, self.size), dtype=numpy.uint)

    def expand(self, x, y):
        if x >= self.size or y >= self.size:
            expand = max([x - self.size + 1, y - self.size + 1])
            self.map = numpy.append(self.map, numpy.zeros((expand, self.size), dtype=numpy.uint), axis=0)
            self.map = numpy.append(self.map, numpy.zeros((self.size + expand, expand), dtype=numpy.uint), axis=1)
            self.size += expand

    def __getitem__(self, coords):
        x, y = coords
        self.expand(x, y)
        return self.map[x, y]

    def __setitem__(self, coords, value):
        x, y = coords
        self.expand(x, y)
        self.map[x, y] = value

    def mark(self, x1, y1, x2, y2):
        if x1 != x2 and y1 != y2:
            return

        self.expand(max([x1, x2]), max([y1, y2]))

        if x1 == x2:
            l = [y1, y2]
            l.sort()
            y1, y2 = l
            for y in range(y1, y2 + 1):
                self[x1, y] += 1

        if y1 == y2:
            l = [x1, x2]
            l.sort()
            x1, x2 = l
            for x in range(x1, x2 + 1):
                self[x, y1] += 1

    def score(self):
        return len([x for x in self.map.flatten().tolist() if x > 1])

    def __str__(self):
        return f"<{self.__class__.__name__}:\nsize: {self.size}\n{self.map.transpose()}\nscore: {self.score()}>"


lines = [line.strip() for line in open('5.input').readlines()]

line_format = re.compile("(?P<x1>[0-9]+),(?P<y1>[0-9]+) -> (?P<x2>[0-9]+),(?P<y2>[0-9]+)")

map = Map()

for line in lines:
    l = line_format.match(line)
    map.mark(int(l['x1']),
             int(l['y1']),
             int(l['x2']),
             int(l['y2']))

print(map.score())

File: 5-1.rb
#!/usr/bin/env ruby

require 'numo/narray'

class Map
  attr_accessor :map

  def initialize
    @size = 1
    @map = Numo::UInt8.zeros(@size, @size)
  end

  def expand(x, y)
    if x >= @size or y >= @size
      expand = [x - @size + 1, y - @size + 1].max
      @map = @map.append(Numo::UInt8.zeros(expand, @size), axis: 0)
      @map = @map.append(Numo::UInt8.zeros(@size + expand, expand), axis: 1)
      @size += expand
    end
  end

  def [](x, y)
    expand(x, y)
    @map[x, y]
  end

  def []=(x, y, val)
    expand(x, y)
    @map[x, y] = val
  end

  def mark(x1, y1, x2, y2)
    return unless x1 == x2 or y1 == y2  # Not orthogonal

    expand([x1, x2].max, [y1, y2].max)

    if x1 == x2
      y1, y2 = [y1, y2].sort
      y1.upto(y2) { |y| self[x1, y] += 1 }
    end

    if y1 == y2
      x1, x2 = [x1, x2].sort
      x1.upto(x2) { |x| self[x, y1] += 1 }
    end
  end

  def score
    Array(@map.flatten).count { |x| x > 1 }
  end

  def to_s
    "<#{self.class}:\nsize: #{@size}\n#{@map.transpose.inspect}\n>"
  end

  def inspect
    "<#{self.class}:\nsize: #{@size}\n#{@map.transpose.inspect}\n>"
  end
end

lines = File.read('5.input').lines.map(&:strip)
line_format = /(?<x1>[[:digit:]]+),(?<y1>[[:digit:]]+) -> (?<x2>[[:digit:]]+),(?<y2>[[:digit:]]+)/

map = Map.new

lines.each do |line|
  line_format.match(line) do |l|
    map.mark(l['x1'].to_i,
             l['y1'].to_i,
             l['x2'].to_i,
             l['y2'].to_i)
  end
end

print map.score, "\n"
